### PR TITLE
Add OBO Credential && Token Exchange Support

### LINKED ISSUE #22363: Feature: Add On-Behalf-Of (OBO) Auth Flow for the Microsoft Graph Team
The [OAuth 2.0 On-Behalf-Of flow (OBO)](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow) serves the use case where an application invokes a service/web API, which in turn needs to call another service/web API. The idea is to propagate the delegated user identity and permissions through the request chain. For the middle-tier service to make authenticated requests to the downstream service, it needs to secure an access token from the Microsoft identity platform, on behalf of the user.

**Related Links:**
Prototype: https://github.com/jongio/azidext/pull/41
Documenation: https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow
.NET Design: https://gist.github.com/christothes/15b9903f7da0954aee166166a8fa1919
MS Graph Team's tracking issue: https://github.com/microsoftgraph/msgraph-sdk-java-core/issues/132

### PR DESCRIPTION
**Pull request by g2vinay**
closes #22363 &&  #22358

### CODE REVIEW

**Review by joshfree — APPROVED** [review]

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/OnBehalfOfCredentialBuilder.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    /**
+     * Configure the User Assertion Scope to be used for OnBehalfOf Authentication request.
+     *
+     * @param userAssertionScope the user assertion to be used for On behalf Of authentication flow
+     * @return An updated instance of this builder with the user assertion scope configured.
+     */
+    public OnBehalfOfCredentialBuilder userAssertionScope(UserAssertionScope userAssertionScope) {
+        this.identityClientOptions.userAssertion(userAssertionScope.getUserAssertion());
+        return this;
+    }
```
Comment (sadasant) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/OnBehalfOfCredentialBuilder.java:39]: oh I like this! I’ll copy this in JS

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/OnBehalfOfCredentialBuilder.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.identity;
+
+public class OnBehalfOfCredentialBuilder extends AadCredentialBuilderBase<OnBehalfOfCredentialBuilder> {
```
Comment (christothes) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/OnBehalfOfCredentialBuilder.java:6]: we'll need to be able to build with a certificate also.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/OnBehalfOfCredentialBuilder.java (hunk 1 above)
Comment (sadasant) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/OnBehalfOfCredentialBuilder.java:42]: this is the advanced OBO credential, right? with a new MSAL client per user assertion?

Comment (g2vinay) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/OnBehalfOfCredentialBuilder.java:95]: this API is outdated and removed now.

**Pull request merged**